walkthrough-w-3

dmosc opened the pull request.

dmosc committed:
dependency (Add) Add Gson dependendy to Maven configuration.

+ Gson dependency allows JSON serialization and deserialization
to easily work with http requests.

*** No breaking changes ***

dmosc committed:
servlet (Comments functionality) Servlet setup to handle in-memory comments registration.

+ Created Comment class to abstract the comment object in the servlet. Comment
class consists of a username and a comment message.

+ Added CommentsServlet.java to manage in-memory comments storage and registration
of new comments. Servlet handles doGet and doPost servlet methods for both operations.

- Removed DataServlet.java since it was not properly describing the functionality
of the comments servlet.

*** No breaking changes ***

dmosc committed:
folder (requests folder) Create new requests folder to store portfolio queries.

+ The folder contains an index.js file that stores queries to the servlets.
The get_comments function fetches the most recent comments submited to the servlet,
and injects HTML elements in a comments list inside the portfolio.

*** No breaking changes ***

dmosc committed:
section (index.html) Add form template to submit comments and comments list to display existing comments.

+ Added to index.html a form to submit new comments to the servlet and an HTML list element
to inject new comments.

+ Added CSS classes needed to override some Bootstrap classes.

- Remove script.js file since it is not being used inside the application.

*** No breaking changes ***

dmosc committed:
dependency (Add Datastore) Add appengine Maven dependency to support DatastoreFactory

New dependency allows the creation of a Datastore to store submitted comments.

*** No breaking changes ***

dmosc committed:
fetaure (Datastore implementation) Substitute in-memory comments storage for App Engine Datastore

+ Added timestamp attribute to Comment.java class to keep track of comments submissions.

+ Modified CommentsServlet.java servlet to store incoming comments into a DatastoreServiceFactory
instead of a private List.

+ Added NewCommentServlet.java to register new comments from the client-side. The servlet registers
the comments in the new Datastore factory.

+ Changed form action from /comments to /comment in index.html to point to the new NewCommentServlet.java
servlet.

*** No breaking changes ***

Domene99 commented on portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:
```diff
@@ -0,0 +1,62 @@
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet that returns some example content. TODO: modify this file to handle comments data */
```
remove

dmosc committed:
config (CODEOWNERS) Add CODEOWNERS file inside .github/

ssmall commented on portfolio/pom.xml:
```diff
@@ -35,7 +45,7 @@
         <version>2.2.0</version>
         <configuration>
           <!-- TODO: set project ID. -->
```
nit: this TODO comment can be removed

dmosc replied:
Will do!

ssmall commented on portfolio/src/main/java/com/google/sps/classes/Comment.java:
```diff
@@ -0,0 +1,25 @@
+package com.google.sps.classes;
+
+public class Comment {
+    private String username;
+    private String comment;
+    private long timestamp;
```
as a best practice, immutable fields should be marked as `final`

ssmall commented on portfolio/src/main/java/com/google/sps/classes/Comment.java:
```diff
@@ -0,0 +1,25 @@
+    private String comment;
+    private long timestamp;
+
+    public Comment(String _username, String _comment, long _timestamp) {
+        username = _username;
+        comment = _comment;
+        timestamp = _timestamp;
+    }
```
nit: prefixing with `_` is not a very common practice in Java. The more idiomatic way to write this constructor in Java would be
```java
    public Comment(String username, String comment, long timestamp) {
        this.username = username;
        this.comment = comment;
        this.timestamp = timestamp;
    }
```

as in [this example](https://github.com/oscardavidrm/step/blob/master/walkthroughs/week-3-server/examples/server-stats/src/main/java/com/google/sps/data/ServerStats.java)

ssmall commented on portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:
```diff
@@ -0,0 +1,62 @@
+
+        Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
+
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
```
Since the instance of the `DatastoreService` can be re-used across requests, you can initialize it in the constructor of `CommentsServlet` and assign it to a `private` field of the class instead of call `DatastoreServiceFactory.getDatastoreService()` for every request.

ssmall commented on portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:
```diff
@@ -0,0 +1,62 @@
+        PreparedQuery comments = datastore.prepare(query);
+
+        List<Comment> comments_to_set = new ArrayList<>();
+        for (Entity entity : comments.asIterable()) {
```
I believe that the call to `.asIterable()` is not necessary; you should be able to write this as

```java
for (Entity entity : comments) {
```

dmosc replied:
It is not reflected in this commit, but I was reading  `.asIterable()` method can receive a `FetchOptions.Builder` object to filter, and/or limit the amount of docs to fetch, which is a feature that is implemented in a following commit.

I was intending to take advantage of this to limit the amount of fetched items. Do you know any other way I could approach this? @ssmall 

Resource: [Javadoc](https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/datastore/FetchOptions)

ssmall replied:
Ah, good to know! Thanks for explaining, that makes sense to leave it like this, then.

ssmall commented on portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:
```diff
@@ -0,0 +1,62 @@
+            String comment = (String) entity.getProperty("comment");
+            long timestamp = (long) entity.getProperty("timestamp");
+
+            Comment comment_to_set = new Comment(username, comment, timestamp);
```
nit: the Google Java Style Guide [recommends](https://google.github.io/styleguide/javaguide.html#s5.2.7-local-variable-names) using lowerCamelCase for local variable names, i.e. `commentToSet`

ssmall commented on portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:
```diff
@@ -0,0 +1,62 @@
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        PreparedQuery comments = datastore.prepare(query);
+
+        List<Comment> comments_to_set = new ArrayList<>();
```
nit: the Google Java Style Guide [recommends](https://google.github.io/styleguide/javaguide.html#s5.2.7-local-variable-names) using lowerCamelCase for local variable names, i.e. `commentsToSet`

ssmall commented on portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:
```diff
@@ -0,0 +1,62 @@
+            comments_to_set.add(comment_to_set);
+        }
+
+        List<Comment> newest_comments = comments_to_set.subList(0, comments_to_set.size() > 5 ? 5 : comments_to_set.size());
```
nit: the Google Java Style Guide [recommends](https://google.github.io/styleguide/javaguide.html#s5.2.7-local-variable-names) using lowerCamelCase for local variable names, i.e. `newestComments`

ssmall replied:
you can use [Math.min](https://docs.oracle.com/javase/8/docs/api/java/lang/Math.html#min-int-int-) here instead of a ternary operator.

ssmall commented on portfolio/src/main/java/com/google/sps/servlets/NewCommentServlet.java:
```diff
@@ -0,0 +1,49 @@
+
+/** Servlet responsible for creating new tasks. */
+@WebServlet("/comment")
+public class NewCommentServlet extends HttpServlet {
```
This class could be combined with `CommentsServlet` into a single class with two methods, `doGet` and `doPost`

ssmall commented on portfolio/src/main/java/com/google/sps/servlets/NewCommentServlet.java:
```diff
@@ -0,0 +1,49 @@
+
+    long timestamp = System.currentTimeMillis();
+    String username, comment;
+    username = request.getParameter("username").length() > 0 ? request.getParameter("username") : "Anonymus";
```
if the `username` parameter is not set for the request, `request.getParameter("username")` will return `null` and this line will throw a `NullPointerException`

ssmall replied:
spelling nit: `"Anonymous"`

anniefu replied:
[nit] Anonymous

ssmall commented on portfolio/src/main/java/com/google/sps/servlets/NewCommentServlet.java:
```diff
@@ -0,0 +1,49 @@
+    long timestamp = System.currentTimeMillis();
+    String username, comment;
+    username = request.getParameter("username").length() > 0 ? request.getParameter("username") : "Anonymus";
+    comment = request.getParameter("comment");
```
What should happen if the `comment` parameter is not set?

ssmall commented on portfolio/src/main/java/com/google/sps/servlets/NewCommentServlet.java:
```diff
@@ -0,0 +1,49 @@
+    username = request.getParameter("username").length() > 0 ? request.getParameter("username") : "Anonymus";
+    comment = request.getParameter("comment");
+
+    Entity comment_entity = new Entity("Comment");
```
same nit as before regarding local variable naming

ssmall commented on portfolio/src/main/webapp/requests/queries/index.js:
```diff
@@ -0,0 +1,24 @@
+const get_comments = async () => {
```
nit: the Google Javascript Style Guide [suggests](https://google.github.io/styleguide/jsguide.html#naming-method-names) that method names should be lowerCamelCased, i.e. `getComments`

ssmall commented on portfolio/src/main/webapp/index.html:
```diff
@@ -7,17 +7,38 @@
                     <div id="social_networks" class="networks"></div>
                 </div>
+                <div class="row">
+                    <form action="/comment" method="POST">
+                        <div class="form-group">
+                            <label for="comment">Tell me something</label>
+                            <textarea name="comment" id="comment" class="form-control" rows="5" placeholder="Comment..."></textarea>
+                        </div>
+                        <div class="input-group mb-2 mr-sm-2">
+                            <div class="input-group-prepend">
+                                <div class="input-group-text">@</div>
+                            </div>
+                            <input type="text" name="username" id="username" class="form-control" placeholder="Username">
+                        </div>
+                        <button type="submit" class="btn btn-primary">Send</button>
+                    </form>
```
Consider marking required inputs as [`required`](https://www.w3schools.com/tags/att_input_required.asp)

ssmall commented on portfolio/src/main/webapp/requests/queries/index.js:
```diff
@@ -0,0 +1,24 @@
+const get_comments = async () => {
+    const payload = await fetch('/comments');
+    const comments = await payload.json();
```
nit: what happens if the server returns an error or the payload is not formatted as JSON?

ssmall commented on portfolio/src/main/webapp/requests/queries/index.js:
```diff
@@ -0,0 +1,24 @@
+    const comments = await payload.json();
+
+    const comments_section = document.getElementById("comments");
+    comments.forEach(comment_payload => {
```
nit: The Google Javascript Style Guide [suggests](https://google.github.io/styleguide/jsguide.html#naming-parameter-names) that parameter names are lowerCamelCased, i.e. `commentPayload`

ssmall commented on portfolio/src/main/webapp/requests/queries/index.js:
```diff
@@ -0,0 +1,24 @@
+        username.classList.add("badge", "badge-dark");
+        comment.innerHTML = comment_payload.comment;
+        
+        const comment_to_append = document.createElement("li");
```
nit: the Google Javascript Style Guide [suggests](https://google.github.io/styleguide/jsguide.html#naming-local-variable-names) that local variable names should be lowerCamelCased, i.e. `commentToAppend`

ssmall reviewed (commented):
Looks great! Mostly style nits and a couple refactorings to consider.

ssmall commented:
Please consider giving this PR a more descriptive title (and description)

anniefu commented on portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:
```diff
@@ -0,0 +1,62 @@
+            comments_to_set.add(comment_to_set);
+        }
+
+        List<Comment> newest_comments = comments_to_set.subList(0, comments_to_set.size() > 5 ? 5 : comments_to_set.size());
```
I'd try cutting off the previous loop after 5 iterations? That way you can save a couple of passes over the ArrayLists.

You could mess around with the for loop statement itself or just keep a separate counter and at the top of the for-loop break out after you hit 5. Something like:

```java
int i = 0;
int num_comments = 5;
for ... {
    if (i >= num_comments) {
        break;
    }
...
    i++;
}
```

I'd actually recommend changing the for-loop statement to break on either the end of comments.asIterable or a counter hitting 5, but I don't know exactly how that'd look :)

dmosc replied:
I actually ended removing this logic since the walk through requested a select option for the user to limit the amount of comments to fetch. But yeah, I believe your approach with the for loop is better! 😄

dmosc committed:
nitfix (Resolve nit observations) Resolve nit observations + walkthrough step 6 integration.

+ Remove unnecessary comments in pom.xml.

+ Changed _ prefix notation in the Comment.java constructor and fixed to a more idiomatic
way "this.attribute" to construct the object. Since attributes are never modified, the final
access modifier was added.

- NewCommentsServlet.java was refactored inside the CommentsServlet.java since both functionalities
were mergeable.

+ CommentsServlet.java now contains the NewCommentsServlet.java doPost method logic, including all the
hotfixes to the nit observations. The query now looks for a Query String limit parameter that can be
attached to limit the amount of comments to fetch from the datastore.

+ The DeleteCommentServlet.java offers the client a delete functionality to remove comments from the
App Engine Datastore. The servlet receives a comment id and tries to remove it.

+ Added required flag to textarea for the comment submission to enforce every comment to include a content.
Moved Network section to the rightmost column since it gives more balance to the page visual load. Hotfixed
function names convention from snake case to camel case.
Added a selection dropdown for the client to limit the amount of comments to load on the comments list.

+ Nitfix on queries/index.js from snake case to camel case convention.

+ Added deleteComment function in mutations/index.js to handle comment deletion triggering by the client.
The function requests the clicked comment to be deleted and removes its node from the list.

+ Added needed classes to style.css to style the new dropdown to limit amount of messages to be loaded.

*** No breaking changes ***

ssmall commented on portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:
```diff
@@ -0,0 +1,96 @@
+
+@WebServlet("/comments")
+public class CommentsServlet extends HttpServlet {
+    private DatastoreService datastore;
```
this field can also be `final` since it is only assigned in the constructor and never changes.

ssmall commented on portfolio/src/main/java/com/google/sps/servlets/DeleteCommentServlet.java:
```diff
@@ -0,0 +1,46 @@
+    public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        response.setContentType("application/json;");
+
+        Key key = KeyFactory.createKey("Comment", Long.parseLong(request.getHeader("id")));
```
what happens if the `id` header is not present?

ssmall commented on portfolio/src/main/webapp/requests/queries/index.js:
```diff
@@ -0,0 +1,42 @@
+        const data = await payload.text();
+        comments = JSON.parse(data); // Verify that the payload is in valid format.
+    } catch (e) {
+        return;
```
nit: it might be nice to do `console.log` with the error here

ssmall reviewed (commented):
Nice! Looking very good. Just a few small comments :)

anniefu commented on portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java:
```diff
@@ -0,0 +1,96 @@
+
+        PreparedQuery comments = datastore.prepare(query);
+        List<Comment> commentsToSet = new ArrayList<>();
+        for (Entity entity : comments.asIterable(FetchOptions.Builder.withLimit(limit))) {
```
nice!

anniefu commented on portfolio/src/main/webapp/requests/queries/index.js:
```diff
@@ -0,0 +1,42 @@
+const getComments = async (limit = 5) => {
```
[nit] It's good practice to make all hard-coded values into variables like `const commentLimit = 5`. 

It's best to always have it available as a variable to avoid another piece of code also hard-coding "5" to match, and then one day someone updates it here, but not in the other piece of code.

dmosc committed:
dependency (Add dependency) Add google gson dependency to support json handling in the server.

*** No breaking changes ***

dmosc committed:
nitfix (Resolve) Resolve nitfixes + refactor portfolio order and view.

+ Resolve mentioned nits, and resolved issue in doPost POST body reading.
Since the body was not being extracted correctly, it was necessary to build
a parser function to extract the body of a request. Added as a return value
in the response of doPost, the new comment created.

+ Added delete functionality to webapp as per requirements. The DeleteCommentServlet.java
expects and id to search for the element to delete from the datastore.

+ Added /functions/index.js to store generic functions to construct HTML elements and
other functional snippets. createCommentElement function creates an element from an
incoming comment.

+ Refactored index.html structure to a more user friendly approach. Removed hobbies section
since it was decided to be removed and changed for a list of projects.

+ Added newCommentForm listener to handle a comment submission in requests/mutations/index.js
The form avoids restarting the webpage and handles the form submission by the client, and sends
it over to the server.

+ Limit variable nitfix + added attributes to getComments() function to add limit and pagination
to the new projects section. The getComments() function fetches a series of my Github repos.

-- Deleted hobbies.js and projects.js since functions were no longer being utilized in the
application. Removed references from /constants/index.js

+ Changed profile picture file.

+ Added some keyFrames and media queries to make the application more smooth and responsive.
Added and refactor servera CSS classes to better fit the new customization.

*** No breaking changes ***
